overlord/registrystate/registrystate.go

```diff
@@ -222,33 +204,310 @@ func updateDatabags(st *state.State, databag registry.JSONDataBag, reg *registry
 	return nil
 }
 
-type cachedRegistryTx struct {
-	account  string
-	registry string
-}
-
+// TODO: none of this applies anymore
 // RegistryTransaction returns the registry.Transaction cached in the context
 // or creates one and caches it, if none existed. The context must be locked by
 // the caller.
-func RegistryTransaction(ctx *hookstate.Context, reg *registry.Registry) (*registry.Transaction, error) {
-	key := cachedRegistryTx{
-		account:  reg.Account,
-		registry: reg.Name,
+//func RegistryTransaction(ctx *hookstate.Context, reg *registry.Registry) (*Transaction, error) {
+//	key := cachedRegistryTx{
+//		account:  reg.Account,
+//		registry: reg.Name,
+//	}
+//	tx, ok := ctx.Cached(key).(*Transaction)
+//	if ok {
+//		return tx, nil
+//	}
+//
+//	var chg *state.Change
+//	if !ctx.IsEphemeral() {
+//		// running in the context of a hook (although not necessarily a registry hook)
+//		task, _ := ctx.Task()
+//
+//		tx, commitTask, err := GetTransaction(task)
+//		if err != nil && !errors.Is(err, &state.NoStateError{}) {
+//			return nil, err
+//		}
+//
+//		if commitTask != nil {
+//			// running in a registry hook, just make sure to save changes once its done
+//			ctx.OnDone(func() error {
+//				commitTask.Set("registry-transaction", tx)
+//				return nil
+//			})
+//
+//			ctx.Cache(key, tx)
+//			return tx, nil
+//		}
+//
+//		// running in a non-registry hook so we'll create the registry hooks/commit
+//		// and add them to the existing change
+//		chg = task.Change()
+//
+//		// TODO: this isn't imposing an order on the registry tasks because I'm not
+//		// sure whether we should chain them on just this one hook or all of them.
+//		// Should all registry changes in the install/configure hooks belong to one
+//		// single transaction? Or is there a reason to have one transaction per hook?
+//	}
+//
+//	// non-hook modification to registry, create tx
+//	st := ctx.State()
+//	tx, err := NewTransaction(st, view)
+//	if err != nil {
+//		return nil, err
+//	}
+//
+//	ctx.OnDone(func() error {
+//		paths := tx.AlteredPaths()
+//		if len(paths) == 0 {
+//			// nothing changed, nothing to commit
+//			return nil
+//		}
+//
+//		if chg == nil {
+//			chg = st.NewChange("commit-registry", fmt.Sprintf("Commit changes to registry \"%s/%s\"", reg.Account, tx.RegistryName))
+//		}
+//
+//		err := populateCommitChange(ctx, chg, tx, reg)
+//		if err != nil {
+//			return err
+//		}
+//
+//		// attach the change ID to the context, so the API knows when the changes
+//		// have been committed
+//		ctx.Cache("change-id", chg.ID())
+//		st.EnsureBefore(0)
+//
+//		return nil
+//	})
+//
+//	ctx.Cache(key, tx)
+//	return tx, nil
+//}
+
+func CreateCommitTasks(ctx *hookstate.Context, tx *Transaction, reg *registry.Registry, plugName string) error {
+	var chg *state.Change
+	if ctx.IsEphemeral() {
+		chg = ctx.State().NewChange("commit-registry", fmt.Sprintf("Commit changes to registry \"%s/%s\"", tx.RegistryAccount, tx.RegistryName))
```

Comment:
the name/description of the change seems a bit off, as it does more than commit?